Multiple enveloped signatures

AndersAbel opened the pull request:
Handle multiple enveloped XML signatures in the same document. This is fairly common for Saml2 implementations.

Fixes #27500

AndersAbel committed:
Fix for handling multiple enveloped xml signatures in the same document.

AndersAbel committed:
Test case for multiple enveloped signatures in document

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs:
```diff
@@ -1637,5 +1637,38 @@ private void ValidateSignedXmlDefaultHashAlgorithms(string expectedSignatureMeth
+            // </root>
+            // Note that signatures were created/added in an order so that all should validate.
+
+            var xml = "<root><x ID=\"a\"><Signature xmlns=\"http://www.w3.org/2000/09/xmldsig#\"><SignedInfo><CanonicalizationMethod Algorithm=\"http://www.w3.org/TR/2001/REC-xml-c14n-20010315\" /><SignatureMethod Algorithm=\"http://www.w3.org/2000/09/xmldsig#rsa-sha1\" /><Reference URI=\"#a\"><Transforms><Transform Algorithm=\"http://www.w3.org/2000/09/xmldsig#enveloped-signature\" /><Transform  ...
```
XML canonicalization should allow the spaces in this to not matter... right?  Can we get it chopped to just look like the comment using a multi-line literal?

```C#
     string xml = @"
<root>
  <x ID=""a"">
  ...
</root>";

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs:
```diff
@@ -1637,5 +1637,38 @@ private void ValidateSignedXmlDefaultHashAlgorithms(string expectedSignatureMeth
+
+            var subject = new SignedXml(doc);
+
+            var parentNode = doc.SelectSingleNode("//x[@ID='" + signatureParent + "']");
```
```suggestion
            IWasGoingToSubstituteTheTypeNameHereButDontKnowItSoClearlyWeCantUseVarHere parentNode = doc.SelectSingleNode("//x[@ID='" + signatureParent + "']");
```

AndersAbel commented:
@bartonjs @krwq Before doing too much work on review, please note that this is still marked as work in progress. I failed to get the complete tests running in my environment, so I pushed to get a CI build. I would be surprised if it passed the build on first try.

danmoseley commented:
@AndersAbel What issue did you have running tests locally? It should work 🙂

AndersAbel committed:
Review fixes

AndersAbel commented:
> @AndersAbel What issue did you have running tests locally? It should work 🙂

@danmoseley I didn't read the instructions on setting up the environment well enough. Once I found my way to https://github.com/dotnet/runtime/blob/main/docs/workflow/building/libraries/README.md it got clearer and it now works.

AndersAbel committed:
Build server troubleshooting
Tests works locally, but not on build server. Temporarily using some dark reflection magic to check
the internal index that the fix is about to see the values as run on the build server.

AndersAbel commented:
I might need some help here on the build and what is expected to pass or not.
The tests failed on the .NET Framework target, so I added some more diagnostics using reflection. That shows that the updated code with the bug fix is not run on .NET framework. Now also the entire test run fails on .NET 7, I don't know if that is due to reflection usage or a temporary issue.

On the .NET Framework target, is the .NET Framework version of SignedXml used? In that case it is expected to fail as that bug was present already in the .NET Framework version of SignedXml. Should I exclude those tests for the .NET Framework target?

I need some feedback on how to proceed.

Once the failure are sorted out I'll remove the reflection again before merge. I can leave them commented out if anyone needs them in future troubleshooting, but I don't think reflection belongs in the live test harness.

krwq commented:
@AndersAbel the easiest way to find out which dll you're using is under the debugger (in VS open the Modules window). Simply put `[SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]` on the test (optionally put an explanation on the attribute with perhaps issue number or some extra info). We already do that in multiple places, i.e.:
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Security.Cryptography.Xml/tests/TransformTest.cs#L82
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Microsoft.CSharp/tests/NullableEnumUnaryOperationTest.cs#L75

AndersAbel committed:
Cleanup before review
- Skip test on .NET Framework target as it will fail.
- Comment out reflection-based tests.
- Add comment with test data generation code for future coders to reference

AndersAbel commented:
Looks like the build failed due to some infrastructure issues and that it is now unavailable. The contents are ready for re-review.

AndersAbel commented:
/azp run

AndersAbel commented:
@krwq @bartonjs Looks like I cannot get the build to rerun. Can you please have a look?

danmoseley commented:
I see it running? 
Btw the trick to force validation to rerun is to close and reopen the PR.

AndersAbel commented:
@danmoseley That's a stale run that has been trying to complete for 2 days... Thanks for the trick, I'll close and reopen.

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/Reference.cs:
```diff
@@ -260,12 +260,20 @@ public void LoadXml(XmlElement value!!)
                         // Hack! this is done to get around the lack of here() function support in XPath
-                        if (transform is XmlDsigEnvelopedSignatureTransform)
+                        if (transform is XmlDsigEnvelopedSignatureTransform
+                            && _uri != null && (_uri.Length == 0 || _uri[0] == '#'))
```
should we throw is `_uri` somehow ends up null?

bartonjs replied:
Null means "whole document"... if I recall correctly?

AndersAbel replied:
No, I don't think so. Null means "custom data source application or first transform should now about". In that case an enveloped signature transform makes no sense and should become a no-op. It is not illegal to include an enveloped signature transform where it is not needed. So with null, I think the reasonable thing to do is to just ignore.
@bartonjs an empty Uri means whole document, and that is taken care of in the code.

krwq commented on src/libraries/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs:
```diff
@@ -1637,5 +1637,97 @@ private void ValidateSignedXmlDefaultHashAlgorithms(string expectedSignatureMeth
+            // To reduce running time, the test data is a pre-calculated string. For anyone that want to
+            // make adjustments to it, this is the small program that was used to generate the data.
+
+            //void Main()
+            //{
+            //	var xml = "<root><x ID=\"a\"/><x ID=\"b\"><x ID=\"c\"><x ID=\"y\"/></x></x></root>";
+            //
+            //	var xd = new XmlDocument();
+            //	xd.LoadXml(xml);
+            //
+            //	Sign(xd, "c", "y");
+            //	Sign(xd, "b", "b");
+            //	Sign(xd, "a", "a");
+            //}
+            //
+            //void Sign(XmlDocument xd, string id, string signaturePlacement)
+            //{
+            //	var sx = new SignedXml(xd);
+            //	var key = RSA.Create();
+            //	sx.SigningKey = key;
+            //
+            //	var reference = new Reference("#" + id);
+            //	reference.AddTransform(new XmlDsigEnvelopedSignatureTransform());
+            //	reference.AddTransform(new XmlDsigExcC14NTransform());
+            //	sx.AddReference(reference);
+            //
+            //	sx.ComputeSignature();
+            //	sx.KeyInfo.AddClause(new RSAKeyValue(key));
+            //
+            //	var node = xd.SelectSingleNode("//x[@ID=\'" + signaturePlacement + "']");
+            //
+            //	var signatureElement = sx.GetXml();
+            //
+            //	node.AppendChild(sx.GetXml());
+            //}
```
Why not calculate it once and assign in a static field?

bartonjs replied:
I think I prefer the approach of using the pregenerated value.  Makes it easier to understand what's going on, and is more like a conformance test.

AndersAbel replied:
I agree with @bartonjs (Because he agrees with me ;) I think that the pregenerated value as it is makes it easier to read the code and understand the test. Including the generation code as a comment is a convenience to any future programmer that want to alter the test case, so they can easily reproduce the current generated data.

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/Reference.cs:
```diff
@@ -260,12 +260,20 @@ public void LoadXml(XmlElement value!!)
+                            XmlNode referenceTarget =
+                                _uri.Length == 0
+                                ? transformElement.OwnerDocument
+                                : SignedXml.GetIdElement(transformElement.OwnerDocument, Utils.GetIdFromLocalUri(_uri, out bool _));
```
we should also test for case where reference is incorrect (i.e. someone tampered with the doc)

bartonjs replied:
True.  While the behavior we're fixing is "we always return failure in this case", let's make sure that we haven't turned complex nesting into so now it "always succeeds, regardless of signature correctness"

AndersAbel replied:
This fix is about the loading of the signature data and not the actual validation. While the test coverage can certainly be improved, I think that is a different task. I do have some validation tests in https://github.com/Sustainsys/Saml2/blob/c466d3bbf12479944e350b649f5b5c825616007b/Tests/Tests.Shared/XmlHelpersTests.cs#L171 that I could carry over. Right now they are testing my wrapper around SignedXml, but essentially they are testing SignedXml.

bartonjs replied:
> This fix is about the loading of the signature data and not the actual validation.

Sure.  But from a black-box perspective we now have a test that shows `CheckSignature()` can return `true` with complex nesting.  Returning `false` when it should is arguably more important than returning `true` when it should, so we should have a test that ensures that this change (from a black box perspective) isn't creating a hole.

Essentially something like load the same document as the positive case, and via a `[Theory]` input change each of a/b/c/y (one per case) to add some extra attribute (or child node, whatever), and show that CheckSignature() returns false.  It doesn't need to care about full matrix testing all of the algorithms, just that we haven't turned complex nesting from "never works" into "free pass".

AndersAbel replied:
Ok, I'll add that.

AndersAbel replied:
Done.

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/Reference.cs:
```diff
@@ -260,12 +260,20 @@ public void LoadXml(XmlElement value!!)
                         // Hack! this is done to get around the lack of here() function support in XPath
-                        if (transform is XmlDsigEnvelopedSignatureTransform)
+                        if (transform is XmlDsigEnvelopedSignatureTransform
+                            && _uri != null && (_uri.Length == 0 || _uri[0] == '#'))
```
make sure to cover case where `_uri` is null (I assume it's when `Uri` attribute is missing)

AndersAbel committed:
Improved tests
 - Test for empty reference URI, which implies whole document

AndersAbel committed:
Improved tests
- Positive and negative tests for tampered data in doc with multiple enveloped signatures

AndersAbel commented:
I've added some more tests and I think I've now handled all the review comments. Please let me know if there is anything more you need to be able to merge this.

bartonjs commented:
Thanks, @AndersAbel

krwq commented on src/libraries/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs:
```diff
@@ -1637,5 +1637,212 @@ private void ValidateSignedXmlDefaultHashAlgorithms(string expectedSignatureMeth
+        public void CheckSignatureDetectsTamperedDataOnMultipleEnvelopedSignatures(
+            string signatureParent, string tamperNode, bool expected)
+        {
+            var tampered = multipleSignaturesXml.Replace($"ID=\"{tamperNode}\"", $"ID=\"{tamperNode}\" Hackerz=\"true\"");
```
I'm still missing a case like `Replace($"ID=\"{tamperNode}\"", $"ID=\"somethingNonExisting\"");` and `Replace($"ID=\"{tamperNode}\"", $"ID=\"somethingElseExisting\"");`

AndersAbel replied:
Added test in #67501 and it actually caught a bug/changed behaviour that I also fixed in that PR.

krwq commented on src/libraries/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs:
```diff
@@ -1637,5 +1637,212 @@ private void ValidateSignedXmlDefaultHashAlgorithms(string expectedSignatureMeth
+        [InlineData("a", "b", true)]
+        [InlineData("b", "b", false)]
+        [InlineData("b", "c", false)]
+        [InlineData("y", "b", true)]
```
why are some of these true?

AndersAbel replied:
Because the altered note is outside the validated content for the signature.
